test/parallel/test-assert-objects.js

```diff
@@ -0,0 +1,384 @@
+  });
+
+  it('should loosely compare two similar objects with type coercion', function() {
+    const obj1 = { a: 1, b: '2' };
+    const obj2 = { a: 1, b: 2 };
+    assert.matchObject(obj1, obj2);
+  });
+
+  it('should strictly compare two objects with different property order', function() {
+    const obj1 = { a: 1, b: 'string' };
+    const obj2 = { b: 'string', a: 1 };
+    assert.matchObjectStrict(obj1, obj2);
+  });
+
+  it('should strictly compare two objects with nested objects', function() {
+    const obj1 = createDeepNestedObject();
+    const obj2 = createDeepNestedObject();
+    assert.matchObjectStrict(obj1, obj2);
+  });
+
+  it('should not strictly compare two objects with different nested objects', function() {
+    const obj1 = createDeepNestedObject();
+    const obj2 = { level1: { level2: { level3: 'differentValue' } } };
+    assert.throws(() => assert.matchObjectStrict(obj1, obj2), Error);
+  });
+
+  it('should loosely compare two objects with nested objects and type coercion', function() {
+    const obj1 = { level1: { level2: { level3: '42' } } };
+    const obj2 = { level1: { level2: { level3: 42 } } };
+    assert.matchObject(obj1, obj2);
+  });
+
+  it('should strictly compare two objects with circular references', function() {
+    const obj1 = createCircularObject();
+    const obj2 = createCircularObject();
+    assert.matchObjectStrict(obj1, obj2);
+  });
+
+  it('should not strictly compare two objects with different circular references', function() {
+    const obj1 = createCircularObject();
+    const obj2 = { self: {} };
+    assert.throws(() => assert.matchObjectStrict(obj1, obj2), Error);
+  });
+
+  it('should loosely compare two objects with circular references', function() {
+    const obj1 = createCircularObject();
+    const obj2 = createCircularObject();
+    assert.matchObject(obj1, obj2);
+  });
+
+  it('should strictly compare two arrays with identical elements', function() {
+    const arr1 = [1, 'two', true];
+    const arr2 = [1, 'two', true];
+    assert.matchObjectStrict(arr1, arr2);
+  });
+
+  it('should not strictly compare two arrays with different elements', function() {
+    const arr1 = [1, 'two', true];
+    const arr2 = [1, 'two', false];
+    assert.throws(() => assert.matchObjectStrict(arr1, arr2), Error);
+  });
+
+  it('should loosely compare two arrays with type coercion', function() {
+    const arr1 = [1, '2', true];
+    const arr2 = [1, 2, 1];
+    assert.matchObject(arr1, arr2);
+  });
+
+  it('should strictly compare two Date objects with the same time', function() {
+    const date1 = new Date('2024-06-10T12:00:00Z');
+    const date2 = new Date('2024-06-10T12:00:00Z');
+    assert.matchObjectStrict(date1, date2);
+  });
+
+  it('should not strictly compare two Date objects with different times', function() {
+    const date1 = new Date('2024-06-10T12:00:00Z');
+    const date2 = new Date('2024-06-11T12:00:00Z');
```

Comment:
nit: we can pass a number to the `Date` constructor, making it obvious which dates are the same, which are different
```suggestion
    const date1 = new Date(0);
    const date2 = new Date(0);
    assert.matchObjectStrict(date1, date2);
  });

  it('should not strictly compare two Date objects with different times', function() {
    const date1 = new Date(0);
    const date2 = new Date(1);
```